docs/history-framework.md

```diff
@@ -193,7 +193,113 @@ sequenceDiagram
+
+##### Pass 1
+First create a map with a type of `{ [entryId] : { entry, nextEntryId }, ...}`.
+Now go through the results. For each entry result:
+- add it to the entry value of map, if there is a nextEntryId don't change it
+- take the current entry's `previousEntryId` look this up in the map and add `nextEntryId` to it with this `previousEntryId` value
+- if there is no `previousEntryId` save this entry as the first entry
+- if the `previousEntryId` already has a different `nextEntryId` it means the history has been corrupted by two or more users writing to history at the same time. Show a warning.
+
+##### Pass 2
+Create a empty array of the history entries.
+Start at the first entry, add it to the array.
+Lookup its nextEntryId in the map. 
+Add this to the array and keep going until you reach an object in the map with no nextEntryId
+To be safe you can use a counter and don't iterate more than the total size of the results. This would prevent infinite loops if there was corrupt data.
+
+##### Storing requirements
+On the start of recording each session we need to know the entry id of the last history entry. So to be safe we have to download all the history entries to find the last one. We could cheat and reverse order the events by the server timestamp and then just look at the last X events. This assumes any out of order entries would be close to each other in time so we don't need to look far back in time to figure this out.
+
+##### Problems
+As described above, to be safe we have to look at every entry at the beginning of a new session with the document in order to figure out what the previousEntryId should be for the first entry of that session.
+
+Possibly this lookup could be done asynchronously. Just the first entry of the session needs to know this, any future entries can refer to this first entry. But still to be safe all of the entries of the last session need to be downloaded. And this could be a large amount of data.
+
+#### Ordering entries based on an index
+
+Store the index of the entry. The first entry will have an index of 0.
+
+To build the correct order we just request the entries from firestore ordered by this index.
+
+##### Storing requirements
+On the start of recording each session we need to know the entry index of the last history entry. This can be found by doing a query with a limit of one sorted by the index in reverse. This would just download the last entry saved.
+
+All entries being written would have to wait for this lookup to occur before they will know their index.
+
+##### Problems
+If multiple users are editing the same document there can be entries with the same index. We can  mitigate this by storing a previousEntryId in each entry. We will know the last entry of the last session because we looked it up to find its index. Even if the sessions start out of order, at some point there will be two entries that have the same previousEntryId (and probably the same index).
```

Comment:
```suggestion
If multiple users are editing the same document there can be entries with the same index. We can mitigate this by storing a `previousEntryId` in each entry. We will know the last entry of the last session because we looked it up to find its index. Even if the sessions start out of order, at some point there will be two entries that have the same previousEntryId (and probably the same index).
```